client.js

```diff
@@ -0,0 +1,145 @@
+        };
+
+        this.vectorFieldSamples = [];
+
+        for (var newSample of newVectorFieldSamples) {
+            var origin = new THREE.Vector3(newSample[0][0], newSample[0][1], newSample[0][2]);
+            var dir = new THREE.Vector3(newSample[1][0], newSample[1][1], newSample[1][2]);
+            var vectorLength = dir.length();
+            // convert the vectorLength to a number between 0 and 1. That way plotted vector length
+            // isn't too large.
+            var strengthFactor = 1 - 1/(1.0 + 100*Math.log(1 + vectorLength))
+            dir.normalize();
+
+            var newArrow = new THREE.ArrowHelper(dir, origin, maxArrowLength * strengthFactor, hex, 
+                maxHeadLength * strengthFactor, maxHeadWidth * strengthFactor);
+            newArrow.cone.material.opacity = strengthFactor;
+            newArrow.line.material.opacity = strengthFactor;
+            newArrow.cone.material.transparent = true;
+            newArrow.line.material.transparent = true;
+
+            this.vectorFieldSamples.push(newArrow);
+            scene.add(newArrow);
+        }
+
+        this.pointCharge.position.x = newChargePosition[0];
+        this.pointCharge.position.y = newChargePosition[1];
+        this.pointCharge.position.z = newChargePosition[2];
+    }
+}
+
+var trajectoryUrl = "http://localhost:5000/trajectory"
+function createFrameBufferUrl(duration, dt) {
+    var url = trajectoryUrl;
+
+    url += "?";
+    url += "dt=" + dt.toString();
+    url += "&";
+    url += "duration=" + duration.toString();
+
+    return url;
+}
+
+class FrameFetcher {
+    constructor(plotter) {
+        this.plotter = plotter;
+
+        this.backFrameBuffer = null;
+        this.frontFrameBuffer = null;
+
+        this.lastPlotTime = null;
+        this.ffwRate = 4.0;
+
+        this.duration = 20.0;
+        this.dt = 0.5;
+    }
+
+    async plotFramesIndefinitely() {
+        this.getNextBuffer();
+
+        while (true) {
+            await this.plotContinuouslyFrontBuffer();
+        }
+    }
+
+    async plotContinuouslyFrontBuffer() {
+        while (this.backFrameBuffer == null) {
+            console.log("waiting for a backFrameBuffer!");
+            await sleep(100);
+        }
+
+        this.frontFrameBuffer = this.backFrameBuffer;
+        this.backFrameBuffer = null;
+
+        this.getNextBuffer();
+
+        for (var frame of this.frontFrameBuffer) {
+            var nextPlotTime = frame[0];
+            if (this.lastPlotTime != null) {
+                await sleep(1000*(nextPlotTime - this.lastPlotTime) / this.ffwRate);
+            }
+            this.lastPlotTime = nextPlotTime;
+
+            this.plotter.updatePlot(frame[2], frame[1]);
+        }
+    }
+
+    setBack(next_frame_buffer) {
+        this.backFrameBuffer = next_frame_buffer;
+    }
+
+    async getNextBuffer() {
+        var frame_buffer = null;
+        var this_ = this;
+        var save_response = async function(response_received) {
```

Comment:
We don't really need async here right? Also, someone told me that we shouldn't create temp variable if we don't need to. Either create a factory function, or better, just plug the function definition directly into get_request below?

Reply:
Yes, I think it would be better to just directly plug in. I made it async because the GET request should be done concurrently while we plot, but it probably doesn't matter because the `http.on_load` doesn't block ?

Reply:
Yeah, `http.on_load` doesn't block. I think a non-async version is probably OK here. Doesn't hurt to leave the `async` though, I think.